packages/remix-react/components.tsx

```diff
@@ -1353,13 +1354,21 @@ type SerializeType<T> = T extends JsonPrimitives
   : T extends (infer U)[]
   ? (U extends NonJsonPrimitives ? null : SerializeType<U>)[]
   : T extends object
-  ? {
-      [k in keyof T as T[k] extends NonJsonPrimitives
-        ? never
-        : k]: SerializeType<T[k]>;
-    }
+  ? SerializeObject<UndefinedOptionals<T>>
   : never;
 
+type SerializeObject<T> = {
+  [k in keyof T as T[k] extends NonJsonPrimitives ? never : k]: SerializeType<
+    T[k]
+  >;
+};
+
+type UndefinedOptionals<T> = {
+  [k in keyof T as undefined extends T[k] ? never : k]: NonNullable<T[k]>;
+} & {
+  [k in keyof T as undefined extends T[k] ? k : never]?: NonNullable<T[k]>;
+};
```

Comment:
Instead of an intersection (which IIRC shows up in intellisense as `{a: A} & {b: B}`, not as `{a: A, b: B}`), would we merge the two mapped objects:

```suggestion
type UndefinedOptionals<T> = Merge<{
  [k in keyof T as undefined extends T[k] ? never : k]: NonNullable<T[k]>;
},{
  [k in keyof T as undefined extends T[k] ? k : never]?: NonNullable<T[k]>;
}>;
```

You can use [`Merge` from `ts-toolbelt`](https://millsp.github.io/ts-toolbelt/modules/object_p_merge.html)

Reply:
Wow, actually really neat that it turns into a clean type. Although this seems like a separate general improvement we could apply to all of the complexly-merged types

Reply:
> Wow, actually really neat that it turns into a clean type. Although this seems like a separate general improvement we could apply to all of the complexly-merged types

True, but we won't switch them all over at once. It more like as we come across `&`s, then we switch them over. So let's start with this one!

Unless you see a good reason not to, of course.

Reply:
I think it would be better to use [`type-fest`](https://github.com/sindresorhus/type-fest)'s [`Merge` type](https://github.com/sindresorhus/type-fest/blob/main/source/merge.d.ts) instead, as that's already one of our dependencies

Reply:
Yeah, I was hesitant to add a new dependency :+1: